addons/product/models/product.py

```diff
@@ -223,9 +223,22 @@ def init(self):
         self.env.cr.execute("CREATE UNIQUE INDEX IF NOT EXISTS product_product_combination_unique ON %s (product_tmpl_id, combination_indices) WHERE active is true"
             % self._table)
 
-    _sql_constraints = [
-        ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
-    ]
+    # TODO: remove before merge
+    # _sql_constraints = [
+    #     ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
+    # ]
+
+    @api.constrains("barcode")
+    def _check_barcode_unique(self):
+        products_with_same_barcode = self.env['product.product'].sudo().search([('barcode', '=', self.barcode)]) - self
```

Comment:
you should make a loop (`self` can be multi records)

```python
for product in self:
      products_with_same_barcode = self.sudo().search([('barcode', '=', product.barcode),('id','!=', product.id])
```

Reply:
Thank you, but this PR is a draft, I just needed to move code around in order to be able to discuss issues related to a task I'm working on. Please do not waste your time reviewing at this stage =)